### DISCUSSION TITLE
How to set multiple languages with babel?

### DISCUSSION THREAD

**Discussion by fredguth** [quarto-dev/quarto-cli#4000] [Q&A] [answered]
It is very common to have academic documents with more than one language.  In latex we would add:

``` partials/in-header.tex
\usepackage[american, brazil]{babel}
```
But this is hardcoded!  I want the possibility of setting different languages with metadata.
How can I create a metadata field in my book project and pass them as options to babel?

```_quarto.yml
#...
babeloptions:
    - american
    - brazil
   ```

**Comment by cscheid** [comment]
If you control the Pandoc template (that is, you're creating new ones), then your template can get information directly from the metadata, like we describe here: https://quarto.org/docs/journals/templates.html#template-partials. That's probably the best option if you're going to use custom LaTeX packages anyway.

**Reply by fredguth** [reply]
@cscheid Is the default `template.tex` this one: https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/formats/pdf/pandoc/template.tex#L249??

This template.tex has (L249):
````
$if(babel-lang)$
\babelprovide[main,import]{$babel-lang$}
$endif$
$for(babel-otherlangs)$
\babelprovide[import]{$babel-otherlangs$}
$endfor$
````

I tried adding in my `_quarto.yml`
```
...
format:
    pdf:
            babel-lang: brazil
            babel-otherlangs: american
````

but it didn't add babel to my generated `bookfilename.tex` file.

I also tried adding to my `in-header.tex`
```
$if(abnt.linguas)$
\PassOptionsToPackage{unicode$for(abnt.linguas)$$abnt.linguas$$sep$,$endfor$}{babel}
$endif$
\usepackage[brazil]{babel}
```
and in `_quarto.yml`
```
...
abnt:
    linguas: american
```

which yields an error: `LaTeX Error: Missing \begin{document}.`

Strangely, if I change `in-header.tex` to:

```
% $if(abnt.linguas)$
% \PassOptionsToPackage{unicode$for(abnt.linguas)$$abnt.linguas$$sep$,$endfor$}{babel}
% $endif$
\usepackage[american,brazil]{babel}
```

it generates a `bookfilename.tex` with the `\usepackage[american,brazil]{babel}` line in it.

**Comment by cderv** [comment, marked as answer]
@fredguth Just to give some context on how Pandoc is handling multiple langue by default - interesting as we are based on Pandoc's template and leveraging such features. 

`lang` in YAML is used to defined main language to use, and any other language needed for other content in a document is to be set on block or inline basis using `lang` attributes on Divs or Spans created.  This is explain in https://pandoc.org/MANUAL.html#language-variables and here is an example 

````markdown
---
format: pdf
lang: en
keep-tex: true
---

This is in English

::: {lang=fr}
Mais cela est écrit en Français
:::

And another [ejemplo]{lang=es}
````
In the generated TeX, we get 
````latex
\ifLuaTeX
\usepackage[bidi=basic]{babel}
\else
\usepackage[bidi=default]{babel}
\fi
\babelprovide[main,import]{english}
\babelprovide[import]{french}
\babelprovide[import]{spanish}
````

You see that mutliple languages are correctly defined for **babel** using `\babelprovide` mechanism (see [babel doc](https://ctan.math.illinois.edu/macros/latex/required/babel/base/babel.pdf))

The specific syntax with Divs or Spans will also make sure that correct LaTeX command is used

````
This is in English

\begin{otherlanguage}{french}

Mais cela est écrit en Français

\end{otherlanguage}

And another \foreignlanguage{spanish}{ejemplo}
````

I believe this mechanism work in Quarto, including Books. 

Anyway, this is how Pandoc supports mutliple languages with babel. 

Did you try this already ? 

Hope it helps

**Reply by fredguth** [reply]
Using `pdf.lang: pt-BR` generates:
````.tex
....
\ifLuaTeX
\usepackage[bidi=basic]{babel}
\else
\usepackage[bidi=default]{babel}
\fi
\babelprovide[main,import]{brazilian}
% get rid of language-specific shorthands (see #6817):
\let\LanguageShortHands\languageshorthands
\def\languageshorthands#1{}
````
But if I add `pdf.otherlangs: [en-US, es-MX]` these other languages are discarded.  I will still try creating a div, but I was looking for a way to set it globally.

**Reply by cderv** [reply]
> But if I add pdf.otherlangs: [en-US, es-MX] these other languages are discarded.

How did you set that ? Is this something Pandoc supports ? are you trying to set globally in YAML ? 

If so, I would say the Pandoc variable used internally when you create Divs is `babel-otherlangs` but that is not documented as mainly for internal usage - so use at your own risk. Setting different language like this globally will not make the LaTeX contains `\begin{otherlanguage}` or `\foreignlanguage` though - that is what the div and Spans are for I think

**Reply by fredguth** [reply]
I included in `format.pdf.babel-otherlangs`, but I will stop doing that and do your way.

**Comment by mcanouil** [comment]
For reference:

- https://github.com/quarto-dev/quarto-cli/discussions/13013
